monkey/monkey_island/cc/repository/mongo_node_repository.py

```diff
@@ -0,0 +1,87 @@
+from copy import deepcopy
+from typing import Any, MutableMapping, Sequence
+
+from pymongo import MongoClient
+
+from monkey_island.cc.models import CommunicationType, MachineID, Node
+
+from . import INodeRepository, RemovalError, RetrievalError, StorageError
+from .consts import MONGO_OBJECT_ID_KEY
+
+UPSERT_ERROR_MESSAGE = "An error occurred while attempting to upsert a node"
+SRC_FIELD_NAME = "machine_id"
+
+
+class MongoNodeRepository(INodeRepository):
+    def __init__(self, mongo_client: MongoClient):
+        self._nodes_collection = mongo_client.monkey_island.nodes
+
+    def upsert_communication(
+        self, src: MachineID, dst: MachineID, communication_type: CommunicationType
+    ):
+        try:
+            node_dict = self._nodes_collection.find_one({SRC_FIELD_NAME: src})
+        except Exception as err:
+            raise StorageError(f"{UPSERT_ERROR_MESSAGE}: {err}")
+
+        if node_dict is None:
+            updated_node = Node(machine_id=src, connections={dst: frozenset((communication_type,))})
+        else:
+            node = MongoNodeRepository._mongo_record_to_node(node_dict)
+            updated_node = MongoNodeRepository._add_connection_to_node(
+                node, dst, communication_type
+            )
+
+        self._upsert_node(updated_node)
+
+    @staticmethod
+    def _mongo_record_to_node(mongo_record: MutableMapping[str, Any]) -> Node:
+        del mongo_record[MONGO_OBJECT_ID_KEY]
+        return Node(**mongo_record)
```

Comment:
This looks applicable to any pydantic object, maybe worth extracting?

Reply:
I had trouble coming up with a good way to do this, but I think I have an idea...